django/db/models/fields/related_descriptors.py

```diff
@@ -125,11 +125,12 @@ def get_prefetch_queryset(self, instances, queryset=None):
         # (related_name ends with a '+'). Refs #21410.
         # The check for len(...) == 1 is a special case that allows the query
         # to be join-less and smaller. Refs #21760.
-        if self.field.remote_field.is_hidden() or len(self.field.foreign_related_fields) == 1:
-            query = {'%s__in' % related_field.name: set(instance_attr(inst)[0] for inst in instances)}
-        else:
-            query = {'%s__in' % self.field.related_query_name(): instances}
-        queryset = queryset.filter(**query)
+        if filter_on_instances:
+            if self.field.remote_field.is_hidden() or len(self.field.foreign_related_fields) == 1:
+                query = {'%s__in' % related_field.name: set(instance_attr(inst)[0] for inst in instances)}
+            else:
+                query = {'%s__in' % self.field.related_query_name(): instances}
+            queryset = queryset.filter(**query)
```

Comment:
I like the use_subquery idea, if it also performs well in practice. There are actually two separate tasks here; getting the items efficiently, and limiting the number of items (see the "pizzas" vs. "vegetarian_pizzas" example in the Prefetch() documentation).